### PR TITLE
451001-ParamonovaAnastasia

### PR DESCRIPTION
**Pull request by AnyaParamonova**

### CODE REVIEW

File: Tracer/Tracer/src/format/ConsoleFormatter.cs [hunk 1]
```diff
@@ -0,0 +1,47 @@
+﻿
+using System.Collections.Generic;
+using MPPTracer.Tree;
+
+namespace MPPTracer.Format
+{
+    public class ConsoleFormatter : Formatter
+    {
+
+        private const string THREAD_TAG = "thread id={0}"+NL;
```
Comment (RamanBut-Husaim) [inline Tracer/Tracer/src/format/ConsoleFormatter.cs]: @AnyaParamonova Please, use `PascalCase` for the constants.

File: Tracer/Tracer/src/Tracer.cs [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+        private Tracer()
+        {
+            rootNode = new RootNode();
+        }
+        public static Tracer Instance
+        {
+            get
+            {
+                if(instance == null)
```
Comment (RamanBut-Husaim) [inline Tracer/Tracer/src/Tracer.cs]: @AnyaParamonova Please, take a look at another possible [singleton implementations](http://csharpindepth.com/Articles/General/Singleton.aspx).

File: Tracer/Tracer/src/format/ConsoleFormatter.cs [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+        private const string METHOD_TAG = "method name={0} time={1}ms class={2} params={3}"+NL;
+
+        public override string Format(TraceResult traceResult)
+        {
+            string result = "";
```
Comment (RamanBut-Husaim) [inline Tracer/Tracer/src/format/ConsoleFormatter.cs]: @AnyaParamonova Use `string.Empty`.

File: Tracer/Tracer/src/format/ConsoleFormatter.cs [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+                result += CreateMethodTree(thread.GetFirstNestedMethod(), 0);
+            }
+
+            return result;
+
+        }
+
+        private string CreateMethodTree(MethodNode rootMethod, int nestingLevel)
+        {
+            string tagList = "";
```
Comment (RamanBut-Husaim) [inline Tracer/Tracer/src/format/ConsoleFormatter.cs]: @AnyaParamonova Use `string.Empty`.

File: Tracer/Tracer/src/format/Formatter.cs [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+
+    public abstract class Formatter : IFormatter
+    {
+        public const string TAB = "\t";
+        public const string NL = "\r\n";
+
+        abstract public string Format(TraceResult traceResult);
+        public string CreateIndent(int nestingLevel)
+        {
+            string indent = "";
```
Comment (RamanBut-Husaim) [inline Tracer/Tracer/src/format/Formatter.cs]: @AnyaParamonova Use `string.Empty`.

Comment (RamanBut-Husaim) [inline Tracer/Tracer/src/format/Formatter.cs]: @AnyaParamonova Try to take a look at string constructor that could take a string value with repetition number.

**Comment by RamanBut-Husaim** [comment]
@AnyaParamonova try to pay more attention to the encapsulation in your code so that many classes and constructors could be `internal` vs `public`.

File: Tracer/Tracer/src/format/ConsoleFormatter.cs [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+        private const string MethodTag = "method name={0} time={1}ms class={2} params={3}";
+
+        public string Format(TraceResult traceResult)
+        {
+            if (traceResult == null)
+            {
+                throw new ArgumentNullException(nameof(traceResult));
+            }
+
+            string result = string.Empty;
```
Comment (RamanBut-Husaim) [inline Tracer/Tracer/src/format/ConsoleFormatter.cs]: @AnyaParamonova Please, use `StringBuilder` in cases when you have to concatenate strings in a loop as strings are immutable in .NET.

File: Tracer/Tracer/src/format/ConsoleFormatter.cs [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+                result += CreateMethodTree(thread.GetEnumerator(), 0)+ NL;
+            }
+
+            return result;
+        }
+
+        private string CreateMethodTree(IEnumerator<MethodNode> enumerator, int nestingLevel)
+        {
+            string indent = new string(Tab, nestingLevel);
+            string tagList = string.Empty;
```
Comment (RamanBut-Husaim) [inline Tracer/Tracer/src/format/ConsoleFormatter.cs]: @AnyaParamonova The same issue there.

File: Tracer/Tracer/src/format/XMLFormatter.cs [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+        }
+
+        private List<XElement> CreateMethodElements(IEnumerator<MethodNode> enumerator)
+        {
+            List<XElement> elements = new List<XElement>();
+            while (enumerator.MoveNext())
+            {
+                MethodNode method = enumerator.Current;
+                MethodDescriptor descriptor = method.Descriptor;
+                object[] attributes = {
```
Comment (RamanBut-Husaim) [inline Tracer/Tracer/src/format/XMLFormatter.cs]: @AnyaParamonova Why `object` array?

File: Tracer/Tracer/src/tree/RootNode.cs [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+        public void StopLastTrace(long endTime)
+        {
+            ThreadNode thread = CurrentThreadNode();
+            thread.StopLastTrace(endTime);
+        }
+
+        private ThreadNode CurrentThreadNode()
+        {
+            int threadId = Thread.CurrentThread.ManagedThreadId;
+            if (ThreadTable.ContainsKey(threadId))
```
Comment (RamanBut-Husaim) [inline Tracer/Tracer/src/tree/RootNode.cs]: @AnyaParamonova Possible thread-safety issue with this check is used outside the lock and other thread could  modify the dictionary a few lines below.

**Comment by RamanBut-Husaim** [comment]
@AnyaParamonova, please leave a comment when you finish all the fixes.

**Comment by AnyaParamonova** [comment]
done

**Pull request merged**